feat(context): Display branch name in mention menu help text

ichim-david opened the pull request:
This commit enhances the mention menu to display the branch name associated with the context item, improving user clarity.

- Adds a `getBranchHelpText` function to determine the appropriate help text based on available branch information.
- The function checks if the branch is specified in the mention data or the query string.
- Updates the MentionMenu component to use the `getBranchHelpText` function to display the branch name.
- Adds unit tests for the `getBranchHelpText` function and `extractRepoAndBranch` function to ensure correct behavior.
- Modifies `extractRepoAndBranch` to support parsing repo and branch from strings like "repo@branch", "repo", or "repo:directory@branch".
- Updates the query to include the branch when searching for directories.


## Test plan

<!-- Required. See https://docs-legacy.sourcegraph.com/dev/background-information/testing_principles. -->

peterguy commented on lib/shared/src/sourcegraph-api/graphql/queries.ts:
```diff
@@ -754,3 +754,27 @@ export const NLS_SEARCH_QUERY = `
+            path
+            byteSize
+            url
+            content
```
Examining the rest of the PR, it looks like you don't need `content`. `content` is pretty expensive, so you should use it only when you need to file contents immediately. If you think you may need the file contents, you can use `rawURL` instead of `content` and then use that URL to retrieve the content.

Additionally, examine whether you need the blob at all - it looks like this PR is using the directories, but not the files, so you may be able to ditch the file info (GitBlob) completely.

julialeex replied:
+1 to the blob comment

ichim-david commented:
@peterguy I'm looking at addressing your feedback.
I noticed that when we perform the remote directories mention we use the `CONTEXT_SEARCH_QUERY_WITH_RANGES` query
https://github.com/sourcegraph/cody/blob/main/lib/shared/src/sourcegraph-api/graphql/queries.ts#L259
which just does a file chunk reply instead of using the entire file.
Ex:
<img width="1672" height="1170" alt="search-context-with-range-results" src="https://github.com/user-attachments/assets/59043675-e8f0-4a76-9387-99369a39a3df" />


If we do just a remote file then we use the entire file as context.
<img width="1684" height="748" alt="Loom Screenshot 2025-07-13 at 03 28 42" src="https://github.com/user-attachments/assets/989318cd-1781-4081-8ed1-538ae0c6280a" />


The new getDirectory with it's conditions would of returned only the files found in that folder and not it's children
<img width="710" height="1122" alt="getDirec" src="https://github.com/user-attachments/assets/2230c00c-a217-43f8-a60c-0dad46c525f3" />
and I think contents help here in order to use the entire file content but I could be wrong.

I know that if I manually add some context I would expect for the entire file to be read instead of just chunks, was trying this on Cursor and although the folder is local and not remote it was able to properly read the contents of all 4 files that were found in that directory.
<img width="1901" height="965" alt="cursor-mention-directories" src="https://github.com/user-attachments/assets/8d2de729-e48e-4b4e-9047-c5b2d55edeb4" />

If we want to keep this light chunk but more encompassing search that finds files also in children directories then I will just look at adding the information regarding the branch searching and selecting and keep the remote directory context with the file chunk as it is.

peterguy commented:
@ichim-david in general, even if you’re using file contents, you should return `rawURL` and make a subsequent call to that URL, rather than adding the burden of the file contents to the HQL query.

julialeex commented:
<img width="278" height="48" alt="Screenshot 2025-07-30 at 2 15 10 PM" src="https://github.com/user-attachments/assets/5acf4f75-0e50-4acc-80c6-38b798d3edc5" />

Can we fix the `%40` and change it to @?

julialeex commented:
> Test that if you select a remote file or directory 

I can't select a remote file

julialeex commented on vscode/src/context/openctx/common/branch-mentions.ts:
```diff
@@ -0,0 +1,220 @@
+
+    const branches = (repoMention.data.branches as string[]) || []
+    const defaultBranch = repoMention.data.defaultBranch as string | undefined
+    const repoId = repoMention.data.repoId as string
```
```suggestion
    const repoId = repoMention.data.repoId as string || ''
```

julialeex commented on vscode/src/context/openctx/common/branch-mentions.ts:
```diff
@@ -0,0 +1,220 @@
+                description: ' ',
+                data: {
+                    repoName,
+                    repoID: repoId,
```
```suggestion
                    repoId: repoId,
```

ichim-david replied:
@julialeex  this is the data structure, repoID not repoId we can't change it, we assign to it repoId that comes from the search

julialeex commented on vscode/src/context/openctx/common/branch-mentions.ts:
```diff
@@ -0,0 +1,220 @@
+            mentions.push({
+                uri: `${serverEndpoint.replace(/\/$/, '')}/${repoName}@${branch}`,
+                title: `@${branch}`,
+                description: ' ',
```
why is this empty with a space?

julialeex commented on vscode/src/context/openctx/common/branch-mentions.ts:
```diff
@@ -0,0 +1,220 @@
+            branchQuery,
+        })
+    } catch (error) {
+        // If the search fails, return empty array to fall back to client-side filtering
```
nit: Can add debug logs here

julialeex commented on vscode/src/context/openctx/common/branch-mentions.ts:
```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    return mentions.filter(isDefined)
```
is this filter necessary here?

julialeex commented on vscode/src/context/openctx/remoteDirectorySearch.ts:
```diff
@@ -2,11 +2,37 @@ import { REMOTE_DIRECTORY_PROVIDER_URI, currentResolvedConfig } from '@sourcegra
+ * Extracts repo name and optional branch from a string.
+ * Supports formats: "repo@branch", "repo", or "repo:directory@branch"
+ */
+export function extractRepoAndBranch(input: string): [string, string | undefined] {
```
`removeFileSearch.ts` has a similar logic for extracting the branch. Can we reuse the code somehow?

julialeex commented on vscode/src/context/openctx/remoteDirectorySearch.ts:
```diff
@@ -58,13 +109,21 @@ async function getDirectoryMentions(repoName: string, directoryPath?: string): P
 
-            const url = `${serverEndpoint.replace(/\/$/, '')}${result.file.url}`
+            // Construct URL with branch information if available
+            const baseUrl = `${serverEndpoint.replace(/\/$/, '')}/${result.repository.name}`
```
`${serverEndpoint.replace(/\/$/, '')}` is also being used at multiple places.

julialeex commented on vscode/src/context/openctx/remoteDirectorySearch.ts:
```diff
@@ -75,33 +134,63 @@ async function getDirectoryMentions(repoName: string, directoryPath?: string): P
-        repoIDs: [repoID],
-        query,
-        filePatterns: [`^${directoryPath}.*`],
+async function getDirectoryBranchMentions(repoName: string, branchQuery?: string): Promise<Mention[]> {
```
ditto. this is similar to getFileBranchMentions

julialeex commented on vscode/src/context/openctx/remoteDirectorySearch.ts:
```diff
@@ -75,33 +134,63 @@ async function getDirectoryMentions(repoName: string, directoryPath?: string): P
+    } = await currentResolvedConfig()
+
+    const items: Item[] = []
+    for (const entry of entries) {
```
We can fetch all contents in parallel instead of sequentially here for better performance.

julialeex reviewed (approved):
Approve to unblock. Thanks David!

peterguy reviewed (approved):
My major concerns about fetching blob contents about were addressed

ichim-david commented:
@thenamankumar You're the only one left :)  feel free to improve this code anyway you seem fit in order to get this merged, now that Julia and Peter have accepted the pr.
Do take a look at the video where I showcases how the flow is now of repo search -> branch select or search -> directory select or search
